Add cvmfs_config fuser command

DrDaveD opened the pull request:
See [CVM-2004](https://sft.its.cern.ch/jira/browse/CVM-2004).  Based on the new `cvmfs_talk device id` command added in  #2745.

DrDaveD committed:
add cvmfs_config fuser command

jblomer reviewed (changes requested):
Looks good to me!  Let's add an integration test for the new command.

DrDaveD committed:
quote the device id just in case

DrDaveD committed:
add test for cvmfs_config fuser

jblomer commented on test/src/094-fuser/main:
```diff
@@ -0,0 +1,45 @@
+  (cd /cvmfs/grid.cern.ch; sleep 30) &
+  backpid="$!"
+  pids="`sudo cvmfs_config fuser grid.cern.ch`"
+  [[ " $pids " == *" $backpid "* ]] || return 10
```
If the test fails, will $backpid survive? If so, we'd need a cleanup.

jblomer commented on test/src/094-fuser/main:
```diff
@@ -0,0 +1,45 @@
+
+  echo "Start background process accessing it and check that it is being used"
+  set -m # enables job control handling
+  (cd /cvmfs/grid.cern.ch; sleep 30) &
```
```suggestion
  (cd /cvmfs/grid.cern.ch; sleep 10000) &
```
I think we should wait essentially infinitely. This should make the test less flaky and in case of a hang the test driver timeout cleans up.

jblomer commented on test/src/094-fuser/main:
```diff
@@ -0,0 +1,45 @@
+  [ -z "$pids" ] || return 11
+
+  echo "Start unshared background process accessing it, and check that"
+  unshare -Um sh -c "cd /cvmfs/grid.cern.ch && /bin/sleep 60" & 
```
```suggestion
  unshare -Um sh -c "cd /cvmfs/grid.cern.ch && /bin/sleep 10000" & 
```
see above

jblomer commented on test/src/094-fuser/main:
```diff
@@ -0,0 +1,45 @@
+  echo "Start unshared background process accessing it, and check that"
+  unshare -Um sh -c "cd /cvmfs/grid.cern.ch && /bin/sleep 60" & 
+  backpid="$!"
+  sleep 1
```
Is this necessary to wait for the unshared process to start?  If so, perhaps the unshared process could create a file and the main test waits for the file to appear. The sleeps have a tendency to fail every now and then (test VMs might be paused).

jblomer commented on test/src/094-fuser/main:
```diff
@@ -0,0 +1,45 @@
+  #   kill -0 on it to see if it is still running
+  if ! kill -0 $backpid; then
+    wait $backpid || true
+    echo "WARNING: could not unshare, skipping unshare test"
```
I think I'd rather like to enable/disable the test based on platforms. Otherwise we risk overlooking the warnings.  We can run it once on all platforms to see where we need to disable it.

jblomer commented on test/src/094-fuser/main:
```diff
@@ -0,0 +1,45 @@
+    CVMFS_GENERAL_WARNING_FLAG=1
+  else
+    pids="`sudo cvmfs_config fuser grid.cern.ch`"
+    [[ " $pids " == *" $backpid "* ]] || return 20
```
As above, let's make sure cleanup is done properly if the test fails

jblomer reviewed (changes requested):
Many thanks!  Some minor comments inline.

We have had a race in the test numbering, could you rename the test to `095-fuser`?

DrDaveD committed:
make improvements to fuser test based on feedback

DrDaveD commented:
I think I addressed all your comments except the one about the warning.  We would could run it as-is to find out which platforms would need to be disabled.  On el7 it would need to depend on whether or or not user namespaces was enabled, I think.  I wonder if the unshare command could be run as root without that.  I would have to test that.

jblomer commented:
@cernvm-bot cloudtest

jblomer commented:
The integration test run looks quite good!

I think it is a good idea to run the unshare command as root.  For the test, it does not matter whether or not user namespaces work. We'd then get EL7 as a test platform.

Only other related failure is a missing `lsof` on the EL8/fuse3 platform. With the new command, we need to add `lsof` as a cvmfs package dependency.

jblomer commented:
Closing for a new PR
